pkg/compactor/block_upload_test.go

```diff
@@ -1591,3 +1813,64 @@ func TestMultitenantCompactor_GetBlockUploadStateHandler(t *testing.T) {
 		})
 	}
 }
+
+// marshalAndUploadJSON is a test helper for uploading a meta file to a certain path in a bucket.
+func marshalAndUploadJSON(t *testing.T, bkt objstore.Bucket, pth string, val interface{}) {
+	t.Helper()
+	err := marshalAndUploadToBucket(context.Background(), bkt, pth, val)
+	require.NoError(t, err)
+}
+
+// populateMetaFileList populates the thanos files list in the meta file based on files in a block directory.
+func populateMetaFileList(t *testing.T, dir string, meta *metadata.Meta) {
+	t.Helper()
+	var files []metadata.File
+	// segment files
+	_, err := os.Stat(filepath.Join(dir, block.ChunksDirname))
+	require.NoError(t, err)
+	err = filepath.Walk(filepath.Join(dir, block.ChunksDirname), func(path string, info os.FileInfo, err error) error {
+		require.NoError(t, err)
+		if info.IsDir() {
+			return nil
+		}
+		files = append(files, metadata.File{
+			RelPath:   filepath.Join(block.ChunksDirname, info.Name()),
+			SizeBytes: info.Size(),
+		})
+		return nil
+	})
+	require.NoError(t, err)
+	// index file
+	stat, err := os.Stat(filepath.Join(dir, block.IndexFilename))
+	require.NoError(t, err)
+	files = append(files, metadata.File{
+		RelPath:   stat.Name(),
+		SizeBytes: stat.Size(),
+	})
+	// meta file
+	files = append(files, metadata.File{
+		RelPath: block.MetaFilename,
+	})
+	meta.Thanos.Files = files
+}
+
+// flipByteAt flips a byte at a given offset in a file.
+func flipByteAt(t *testing.T, fname string, offset int64) {
+	t.Helper()
+	fd, err := os.OpenFile(fname, os.O_RDWR, 0644)
+	defer func(fd *os.File) {
```

Comment:
We only need to `defer` if there was no error. If there was error, `fd` will be nic, and `fd.Close()` in defer will return error, which will then be reported, instead of original (and more important) error from `OpenFile`.